libsql-wal/src/bottomless/job.rs

```diff
@@ -0,0 +1,442 @@
+}
+
+#[derive(Debug, AsBytes, FromZeroes, FromBytes)]
+#[repr(C)]
+pub struct CompactedSegmentDataFooter {
+    checksum: lu32,
+}
+
+#[repr(transparent)]
+struct BytesLike<T>(pub T);
+
+impl<T> AsRef<[u8]> for BytesLike<T>
+where
+    T: AsBytes,
+{
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl<C, F> Job<C, Arc<SealedSegment<F>>>
+where
+    F: FileExt,
+{
+    /// Perform the job and return the JobResult. This is not allowed to panic.
+    pub(crate) async fn perform<S, IO>(
+        self,
+        storage: S,
+        io: IO,
+    ) -> JobResult<C, Arc<SealedSegment<F>>>
+    where
+        S: Storage<Config = C>,
+        IO: Io,
+    {
+        let result = self.try_perform(storage, io).await;
+        JobResult { job: self, result }
+    }
+
+    async fn try_perform<S, IO>(&self, storage: S, io: IO) -> Result<u64>
+    where
+        S: Storage<Config = C>,
+        IO: Io,
+    {
+        let segment = &self.request.segment;
+        let segment_id = io.uuid();
+        let tmp = io.tempfile()?;
+
+        let new_index = compact(segment, &tmp, segment_id).await?;
+
+        let meta = SegmentMeta {
+            segment_id,
+            namespace: self.request.namespace.clone(),
+            start_frame_no: segment.header().start_frame_no.get(),
+            end_frame_no: segment.header().last_committed(),
+            created_at: io.now(),
+        };
+        let config = self
+            .request
+            .storage_config_override
+            .clone()
+            .unwrap_or_else(|| storage.default_config());
+
+        storage.store(&config, meta, tmp, new_index).await?;
+
+        Ok(segment.header().last_committed())
+    }
+}
+
+/// Compact a sealed segment into out-file with id `segment_id`, and returns the new index.
+// todo: move to segment module
+async fn compact(
+    segment: &SealedSegment<impl FileExt>,
+    out_file: &impl FileExt,
+    segment_id: uuid::Uuid,
+) -> Result<Vec<u8>> {
+    let mut hasher = crc32fast::Hasher::new();
+
+    let header = CompactedSegmentDataHeader {
+        frame_count: (segment.index().len() as u64).into(),
+        segment_id: segment_id.as_u128().into(),
+        start_frame_no: segment.header().start_frame_no,
+        end_frame_no: segment.header().last_commited_frame_no,
+    };
+
+    hasher.update(header.as_bytes());
+    let (_, ret) = out_file
+        .write_all_at_async(ZeroCopyBuf::new_init(header), 0)
+        .await;
+    ret?;
+
+    let mut pages = segment.index().stream();
+    // todo: use Frame::Zeroed somehow, so that header is aligned?
+    let mut buffer = Box::new(ZeroCopyBuf::<Frame>::new_uninit());
+    let mut out_index = fst::MapBuilder::memory();
+    let mut current_offset = 0;
+
+    while let Some((page_no_bytes, offset)) = pages.next() {
+        let page_no = u32::from_be_bytes(page_no_bytes.try_into().unwrap());
+        let (b, ret) = segment.read_frame_offset_async(offset as _, buffer).await;
+        ret.unwrap();
```

Comment:
whats this unwrap for?

Reply:
Uh yeah, I should return an io error here, I'll do in a follow up